packages/smart-contracts/src/contracts/BatchPayments.sol

```diff
@@ -0,0 +1,178 @@
+     * @param _recipients Takes a list of accounts.
+     * @param _amounts Takes a list of amounts, corresponding to recipients[].
+     * @dev Remember to give approval for this contract to spend user tokens.
+     */
+    function batchOrphanERC20Payments(
+        IERC20 _token, 
+        address[] calldata _recipients, 
+        uint256[] calldata _amounts
+    ) external {
+        for (uint256 i = 0; i < _recipients.length; i++)
+            _token.transferFrom(msg.sender, _recipients[i], _amounts[i]);
+    }
+
+    
+    /**
+    * @notice Send a batch of Ethereum payments w/fees with paymentReferences to multiple accounts.
+    *         The sum of _amounts and _feeAMounts must be <= to msg.value.
+    *         If one payment failed, the whole batch is reverted
+    * @param _recipients List of recipients accounts.
+    * @param _amounts List of amounts, corresponding to recipients[].
+    * @param _paymentReferences List of paymentRefs, corr. to the recipients[].
+    * @param _feeAmounts List of amounts of the payment fee, corr. to the recipients[].
+    * @param _feeAddress The fee recipient.
+    * @dev Uses EthereumFeeProxy.sol to pay an invoice and fees, with a payment reference.
+     */
+    function batchEthereumPaymentsWithReferenceAndFee(
+        address[] calldata _recipients, 
+        uint256[] calldata _amounts,
+        bytes[] calldata _paymentReferences,
+        uint256[] calldata _feeAmounts,
+        address _feeAddress 
+    ) external payable nonReentrant {
+        // EOA transfer token on the contract
+        payable(address(this)).transfer(msg.value);
+        uint256 toReturn = msg.value;
+
+        // Contract pays the batch payment
+        for (uint256 i = 0; i < _recipients.length; i++) {
+            toReturn -= (_amounts[i]+_feeAmounts[i]);
+            paymentEthereumFeeProxy.transferWithReferenceAndFee{value: _amounts[i]+_feeAmounts[i]}(
+                payable(_recipients[i]), 
+                _paymentReferences[i],
+                _feeAmounts[i],
+                payable(_feeAddress)
+            );
+        }
+
+        // Transfer the remaining ethers to the sender
+        (bool sendBackSuccess, ) = payable(msg.sender).call{ value: toReturn }('');
+        require(sendBackSuccess, 'Could not send remaining funds to the payer');
+    }
+
+    /**
+     * @notice Send a batch of erc20 payments w/fees with paymentReferences to multiple accounts.
+     * @param _tokenAddress Token to transact with.
+     * @param _recipients List of recipients accounts.
+     * @param _amounts List of amounts, corresponding to recipients[].
+     * @param _paymentReferences List of paymentRefs, corr. to the recipients[] and .
+     * @param _feeAmounts List of amounts of the payment fee, corr. to the recipients[].
+     * @param _feeAddress The fee recipient.
+     * @dev Uses ERC20FeeProxy.sol to pay an invoice and fees, with a payment reference.
+     */
+    function batchERC20PaymentsWithReferenceAndFee(
```

Comment:
what's the reason for this method? does it reduce the gas usage?

Reply:
Kept it for gas economy.